source/grammar/qasm3.g4

```diff
@@ -276,34 +279,73 @@ logicalOperator
     | '||'
     ;
 
+shiftOperator
+    : '<<'
+    | '>>'
+    ;
+
+additiveOperator
+    : '+'
+    | '-'
+    ;
+
+multiplicativeOperator
+    : '*'
+    | '/'
+    | '%'
+    ;
+
+powerOperator
+    : '**'
+    ;
+
 expressionStatement
     : expression SEMICOLON
     ;
 
 expression
     // include terminator/unary as base cases to simplify parsing
     : expressionTerminator
-    | unaryExpression
     // expression hierarchy
-    | xOrExpression
-    | expression '|' xOrExpression
+    | booleanExpression
     ;
 
 /**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
   *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
   * Order (first to last evaluation):
     Terminator (including Parens),
     Unary Op,
+    Power
     Multiplicative
     Additive
     Bit Shift
     Bit And
     Exlusive Or (xOr)
     Bit Or
+    Comparison
```

Comment:
I'd just replace these w/ boolean expression

Reply:
Original `booleanExpression` is not correct. LogicalAND (&&) must be prioritized than LogicalOR (||) if QASM3 follows C.

I do not think `membershipTest` is necessary in the hierarchy because `in` is a kind of `logicalOperator` in other language such as Python. If we introduce an array types, `membershipTest` will become `Identifier 'in' orExpression`, which is mostly same with `comparisonExpression`: `comparisonExpression relationalOperator orExpression`. In this PR, I proposed `relationalOperator` includes `in` and simplified the hierarchy by merging `membershipTest` into `comparisonExpression`.

Original `booleanExpression`:
```
booleanExpression
    : membershipTest
    | comparisonExpression
    | booleanExpression logicalOperator comparisonExpression
    ;

comparisonExpression
    : expression  // if (expression)
    | expression relationalOperator expression
    ;
```

Reply:
Fair pt